presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,48 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        if (preGroupedHashStrategy.isPresent()) {
+            // Running in segmented aggregation mode.
+            int lastRowInPage = page.getPositionCount() - 1;
+            if (firstUnfinishedSegment == null) {
+                // If this is the first page, treat the first segment in this page as the current segment.
+                firstUnfinishedSegment = page.getRegion(0, 1);
+            }
+
+            if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), lastRowInPage, page.extractChannels(preGroupedChannels))) {
+                // If the current segment might have more data in the incoming pages, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
+            }
+            else if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
+                // If the current page starts with a new segment, flush before processing it.
+                remainingPageForSegmentedAggregation = page;
+            }
+            else {
+                // If the current segment ends in the current page, flush it with all the segments (if exist) except the last segment of the current page.
+                int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
+                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
+                remainingPageForSegmentedAggregation = page.getRegion(lastSegmentStart, lastRowInPage - lastSegmentStart + 1);
+            }
+            // Record the last segment.
+            firstUnfinishedSegment = page.getRegion(lastRowInPage, 1);
```

Comment:
Seems that lastRowInPage is never updated. So this  **firstUnfinishedSegment** is always the last row in the page? What does it mean? Could you add member field comments for the new fields added in this PR?